solidity/contracts/vault/TBTCOptimisticMinting.sol

```diff
@@ -0,0 +1,316 @@
+        //       decide to cut fees here, it is possible to fix this imbalance
+        //       later by a donation to the Bridge (see DonationVault).
+
+        _mint(deposit.depositor, deposit.amount);
+        optimisticMintingDebt[deposit.depositor] += deposit.amount;
+
+        delete pendingOptimisticMints[depositKey];
+
+        emit OptimisticMintingFinalized(
+            msg.sender,
+            deposit.depositor,
+            deposit.amount,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    /// @notice Allows a guard to cancel optimistic minting request. The
+    ///         following conditions must be met:
+    ///         - The optimistic minting request for the given deposit exists.
+    ///         - The optimistic minting request for the given deposit has not
+    ///           been finalized yet.
+    function cancelOptimisticMint(
+        bytes32 fundingTxHash,
+        uint32 fundingOutputIndex
+    ) external onlyGuard {
+        uint256 depositKey = calculateDepositKey(
+            fundingTxHash,
+            fundingOutputIndex
+        );
+
+        require(
+            pendingOptimisticMints[depositKey] > 0,
+            "Optimistic minting not requested of already finalized"
+        );
+
+        delete pendingOptimisticMints[depositKey];
+
+        emit OptimisticMintingCancelled(
+            msg.sender,
+            fundingTxHash,
+            fundingOutputIndex,
+            depositKey
+        );
+    }
+
+    // TODO: Find a convenient way for a guard to block minting at 3AM and deal
+    //       with an errant minter.
+
+    /// @notice Adds the address to the minter set.
+    function addMinter(address minter) external onlyOwner {
+        require(!isMinter[minter], "This address is already a minter");
+        isMinter[minter] = true;
+        emit MinterAdded(minter);
+    }
+
+    /// @notice Removes the address from the minter set.
+    function removeMinter(address minter) external onlyOwner {
+        require(isMinter[minter], "This address is not a minter");
+        delete isMinter[minter];
+        emit MinterRemoved(minter);
+    }
+
+    /// @notice Adds the address to the guard set.
+    function addGuard(address guard) external onlyOwner {
+        require(!isGuard[guard], "This address is already a guard");
+        isGuard[guard] = true;
+        emit GuardAdded(guard);
+    }
+
+    /// @notice Removes the address from the guard set.
+    function removeGuard(address guard) external onlyOwner {
+        require(isGuard[guard], "This address is not a guard");
+        delete isGuard[guard];
+        emit GuardRemoved(guard);
+    }
+
+    /// @notice Calculates deposit key the same way as the Bridge contract.
+    ///         The deposit key is computed as
+    ///         keccak256(fundingTxHash | fundingOutputIndex).
+    function calculateDepositKey(
```

Comment:
Since we have this public function that can be used by anyone, do you think we can pass only `depositKey` to `optimisticMint`, `finalizeOptimisticMint`, and `cancelOptimisticMint()`?

Reply:
We are passing funding TX hash and funding output index on purpose - this lets to capture them on-chain. Guardian needs only Etherscan and Blockstream.info to see if the given deposit happened and how many confirmations it has.